From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hippietrail <hippytrail@gmail.com>
Date: Wed, 30 Oct 2024 06:06:30 +0700
Subject: [PATCH] 7126: M68k a line trap support for at least Classic Mac,
 Apple Lisa, and Sharp X68000

mac-style and lisa-style a-line traps for m68k

@elif

add sharp x68000 f-line trap support
---
 .../Processors/68000/certification.manifest   |  3 +
 .../68000/data/languages/68000.ldefs          | 61 +++++++++++++++++++
 .../68000/data/languages/68000.sinc           | 10 +++
 .../68000/data/languages/lisatrap.slaspec     |  2 +
 .../68000/data/languages/mactrap.slaspec      |  2 +
 .../68000/data/languages/x68ktrap.slaspec     |  2 +
 6 files changed, 80 insertions(+)
 create mode 100644 Ghidra/Processors/68000/data/languages/lisatrap.slaspec
 create mode 100644 Ghidra/Processors/68000/data/languages/mactrap.slaspec
 create mode 100644 Ghidra/Processors/68000/data/languages/x68ktrap.slaspec

diff --git a/Ghidra/Processors/68000/certification.manifest b/Ghidra/Processors/68000/certification.manifest
index caf51979e7..b729211fb8 100644
--- a/Ghidra/Processors/68000/certification.manifest
+++ b/Ghidra/Processors/68000/certification.manifest
@@ -13,6 +13,9 @@ data/languages/68030.slaspec||GHIDRA||||END|
 data/languages/68040.slaspec||GHIDRA||||END|
 data/languages/coldfire.slaspec||GHIDRA||||END|
 data/languages/CPU32.slaspec||GHIDRA||||END|
+data/languages/lisatrap.slaspec||GHIDRA||||END|
+data/languages/mactrap.slaspec||GHIDRA||||END|
+data/languages/x68ktrap.slaspec||GHIDRA||||END|
 data/manuals/68000.idx||GHIDRA||||END|
 data/patterns/68000_patterns.xml||GHIDRA||||END|
 data/patterns/patternconstraints.xml||GHIDRA||||END|
diff --git a/Ghidra/Processors/68000/data/languages/68000.ldefs b/Ghidra/Processors/68000/data/languages/68000.ldefs
index 4b9135c686..1805c810ec 100644
--- a/Ghidra/Processors/68000/data/languages/68000.ldefs
+++ b/Ghidra/Processors/68000/data/languages/68000.ldefs
@@ -86,4 +86,65 @@
     <compiler name="default" spec="68000.cspec" id="default"/>
     <external_name tool="IDA-PRO" name="CPU32"/>
   </language>
+
+  <language processor="68000"
+            endian="big"
+            size="32"
+            variant="MacTrap"
+            version="1.1"
+            slafile="mactrap.sla"
+            processorspec="68000.pspec"
+            manualindexfile="../manuals/68000.idx"
+            id="68000:BE:32:MacTrap">
+    <description>Motorola 32-bit MacTrap</description>
+    <compiler name="default" spec="68000.cspec" id="default"/>
+    <compiler name="register" spec="68000_register.cspec" id="register"/>
+    <external_name tool="gnu" name="m68k"/>
+    <external_name tool="IDA-PRO" name="68000"/>
+    <external_name tool="IDA-PRO" name="68040"/>
+    <external_name tool="IDA-PRO" name="68K"/>
+    <external_name tool="DWARF.register.mapping.file" name="68000.dwarf"/>
+    <external_name tool="qemu" name="qemu-m68k"/>
+  </language>
+
+  <language processor="68000"
+            endian="big"
+            size="32"
+            variant="LisaTrap"
+            version="1.1"
+            slafile="lisatrap.sla"
+            processorspec="68000.pspec"
+            manualindexfile="../manuals/68000.idx"
+            id="68000:BE:32:LisaTrap">
+    <description>Motorola 32-bit LisaTrap</description>
+    <compiler name="default" spec="68000.cspec" id="default"/>
+    <compiler name="register" spec="68000_register.cspec" id="register"/>
+    <external_name tool="gnu" name="m68k"/>
+    <external_name tool="IDA-PRO" name="68000"/>
+    <external_name tool="IDA-PRO" name="68040"/>
+    <external_name tool="IDA-PRO" name="68K"/>
+    <external_name tool="DWARF.register.mapping.file" name="68000.dwarf"/>
+    <external_name tool="qemu" name="qemu-m68k"/>
+  </language>
+
+  <language processor="68000"
+            endian="big"
+            size="32"
+            variant="X68KTrap"
+            version="1.1"
+            slafile="x68ktrap.sla"
+            processorspec="68000.pspec"
+            manualindexfile="../manuals/68000.idx"
+            id="68000:BE:32:X68KTrap">
+    <description>Motorola 32-bit X68KTrap</description>
+    <compiler name="default" spec="68000.cspec" id="default"/>
+    <compiler name="register" spec="68000_register.cspec" id="register"/>
+    <external_name tool="gnu" name="m68k"/>
+    <external_name tool="IDA-PRO" name="68000"/>
+    <external_name tool="IDA-PRO" name="68040"/>
+    <external_name tool="IDA-PRO" name="68K"/>
+    <external_name tool="DWARF.register.mapping.file" name="68000.dwarf"/>
+    <external_name tool="qemu" name="qemu-m68k"/>
+  </language>
+
 </language_definitions>
diff --git a/Ghidra/Processors/68000/data/languages/68000.sinc b/Ghidra/Processors/68000/data/languages/68000.sinc
index 2da450d71f..2a46fbdd8c 100644
--- a/Ghidra/Processors/68000/data/languages/68000.sinc
+++ b/Ghidra/Processors/68000/data/languages/68000.sinc
@@ -89,7 +89,9 @@ define token instr (16)
   op01 = (0,1)
   op02 = (0,2)
   op03 = (0,3)
+  op07 = (0,7)
   op08 = (0,8)
+  op011 = (0,11)
   op015 = (0,15)
   op34 = (3,4)
   op35 = (3,5)
@@ -925,6 +927,14 @@ macro rotateRightExtended(count, register, width) {
 
 with : extGUARD=1 {
 
+@ifdef MACTRAP
+:atrapm "#"op011 is op=10 & op011 unimpl
+@elif defined(LISATRAP)
+:atrapl "#"t is op=10 & op011; opx015 [ t = op011 * 65536 + opx015; ] unimpl
+@elif defined(X68KTRAP)
+:ftrap "#"op07 is opbig=255 & op07 unimpl
+@endif # X68KTRAP
+
 :abcd Tyb,Txb			is op=12 & op48=16 & Tyb & Txb {
 	CF = carry(Txb,carry(Tyb,XF));
 	Txb = Txb + Tyb + XF;
diff --git a/Ghidra/Processors/68000/data/languages/lisatrap.slaspec b/Ghidra/Processors/68000/data/languages/lisatrap.slaspec
new file mode 100644
index 0000000000..d05fb9dee9
--- /dev/null
+++ b/Ghidra/Processors/68000/data/languages/lisatrap.slaspec
@@ -0,0 +1,2 @@
+@define LISATRAP ""
+@include "68000.sinc"
diff --git a/Ghidra/Processors/68000/data/languages/mactrap.slaspec b/Ghidra/Processors/68000/data/languages/mactrap.slaspec
new file mode 100644
index 0000000000..7e0856dd79
--- /dev/null
+++ b/Ghidra/Processors/68000/data/languages/mactrap.slaspec
@@ -0,0 +1,2 @@
+@define MACTRAP ""
+@include "68000.sinc"
diff --git a/Ghidra/Processors/68000/data/languages/x68ktrap.slaspec b/Ghidra/Processors/68000/data/languages/x68ktrap.slaspec
new file mode 100644
index 0000000000..ac8c913719
--- /dev/null
+++ b/Ghidra/Processors/68000/data/languages/x68ktrap.slaspec
@@ -0,0 +1,2 @@
+@define X68KTRAP ""
+@include "68000.sinc"
-- 
2.45.1

